pkg/tfgen/namecheck.go

```diff
@@ -17,150 +17,341 @@ package tfgen
+		ps := nameCheckPropSet{
+			location:  loc,
+			props:     props[tokens.Token(tok)],
+			schemaMap: prov.P.DataSourcesMap().Get(key).Schema(),
+			infos:     prov.DataSources[key].Fields,
 		}
-		if ne[i].message < ne[j].message {
-			return true
+		all = append(all, ps)
+	}
+
+	for tok := range spec.Types {
+		typePaths := renamesBuilder.ObjectTypePaths(tokens.Type(tok))
+		for _, typePath := range typePaths.Paths() {
+			loc := fmt.Sprintf("Type: %q\n    TypePath: %s", tok, typePath.String())
+			p, err := findPath(prov, typePath)
+			if err != nil {
+				return err
+			}
+			schemaMap, err := p.SchemaMap()
+			if err != nil {
+				return err
+			}
+			infos, err := p.SchemaInfos()
+			if err != nil {
+				return err
+			}
+			ps := nameCheckPropSet{
+				location:  loc,
+				props:     props[tokens.Token(tok)],
+				schemaMap: schemaMap,
+				infos:     infos,
+			}
+			all = append(all, ps)
 		}
-		return false
-	})
+	}
+
+	for _, props := range all {
+		nameCheckUniquePulumiNames(sink, props)
+		nameCheckUniqueTerraformNames(sink, props)
+		nameCheckTerraformToPulumiName(sink, props)
+		nameCheckPulumiToTerraformName(sink, props)
+	}
+
+	return nil
 }
 
-func (ne nameErrors) Report(sink diag.Sink) {
-	if len(ne) == 0 {
-		return
+func nameCheckTerraformToPulumiName(sink diag.Sink, props nameCheckPropSet) {
+	for _, p := range props.props {
+		expected := p.Name
+		actual := tfbridge.TerraformToPulumiNameV2(p.Key, props.schemaMap, props.infos)
+		if actual == expected.String() {
+			continue
+		}
+		m := "TerraformToPulumiNameV2(%q) unexpectedly returns %q, expecting %q from schema generation.\n" +
+			"  Please report this as a bug to pulumi/pulumi-terraform-bridge.\n" +
+			"  Location:\n    %v\n"
+		sink.Warningf(&diag.Diag{Message: m}, p.Key, actual, expected, props.location)
 	}
+}
 
-	if sink == nil {
-		sink = diag.DefaultSink(os.Stdout, os.Stderr, diag.FormatOptions{
-			Color: colors.Always,
-		})
+func nameCheckPulumiToTerraformName(sink diag.Sink, props nameCheckPropSet) {
+	for _, p := range props.props {
+		expected := p.Key
+		actual := tfbridge.PulumiToTerraformName(p.Name.String(), props.schemaMap, props.infos)
+		if actual == expected {
+			continue
+		}
+		m := "PulumiToTerraformName(%q) unexpectedly returns %q, expecting %q from schema generation.\n" +
+			"  Please report this as a bug to pulumi/pulumi-terraform-bridge.\n" +
+			"  Location:\n    %v\n"
+		sink.Warningf(&diag.Diag{Message: m}, p.Key, actual, expected, props.location)
 	}
+}
+
+func nameCheckUniquePulumiNames(sink diag.Sink, props nameCheckPropSet) {
+	index := map[tokens.Name][]string{}
+	for _, p := range props.props {
+		index[p.Name] = append(index[p.Name], p.Key)
+	}
+	for k, v := range index {
+		if len(v) <= 1 {
+			continue
+		}
+		m := "Pulumi property %q maps to multiple Terraform properties %v.\n" +
+			"  This will cause runtime errors in the provider.\n" +
+			"  Please set SchemaInfo.Name to rename the Terraform properties.\n" +
```

Comment:
@danielrbradley as we discussed the warnings should include some steps to remediation; if this ever pops up again, this is what it would say - does this make sense? Or can I change the wording further. Full paths to SchemaInfo can be involved, e.g. ProviderInfo.Resources[t].Fields.Elem....Name